Dockerfile

```diff
@@ -1,108 +1,193 @@
+
+# Set up a Python virtual environment.
+# It is already 'activated' because $VIRTUAL_ENV/bin was put on $PATH.
+RUN python3.8 -m venv "$VIRTUAL_ENV"
 
-# Install Python requirements
-COPY setup.py setup.py
-COPY common common
-COPY openedx openedx
-COPY lms lms
-COPY cms cms
-COPY requirements/pip.txt requirements/pip.txt
-COPY requirements/edx/base.txt requirements/edx/base.txt
+# Install Python requirements.
+# Requires copying over requirements files, but not entire repository.
+# We filter out the local ('common/*' and 'openedx/*', and '.') Python projects,
+# because those require code in order to be installed. They will be installed
+# later. This step can be simplified when the local projects are dissolved
+# (see https://openedx.atlassian.net/browse/BOM-2579).
+COPY requirements requirements
+RUN  sed '/^-e \(common\/\|openedx\/\|.\)/d' requirements/edx/base.txt \
+  > requirements/edx/base-minus-local.txt
 RUN pip install -r requirements/pip.txt
-RUN pip install -r requirements/edx/base.txt
+RUN pip install -r requirements/edx/base-minus-local.txt
 
-# Copy just JS requirements and install them.
+# Set up a Node environment and install Node requirements.
+# Must be done after Python requirements, since nodeenv is installed
+# via pip.
+# The node environment is already 'activated' because its .../bin was put on $PATH.
+RUN nodeenv /edx/app/edxapp/nodeenv --node=12.11.1 --prebuilt
 COPY package.json package.json
 COPY package-lock.json package-lock.json
-RUN nodeenv /edx/app/edxapp/nodeenv --node=12.11.1 --prebuilt
 RUN npm set progress=false && npm install
 
-ENV LMS_CFG /edx/etc/lms.yml
-ENV STUDIO_CFG /edx/etc/studio.yml
-
-# Copy over remaining code.
-# We do this as late as possible so that small changes to the repo don't bust
-# the requirements cache.
+# Copy over remaining parts of repository (including all code).
 COPY . .
 
+# Install Python requirements again in order to capture local projects, which
+# were skipped earlier. This should be much quicker than if were installing
+# all requirements from scratch.
+RUN pip install -r requirements/edx/base.txt
+
+
+##################################################
+# Define LMS non-dev target.
 FROM base as lms
 ENV SERVICE_VARIANT lms
-ENV DJANGO_SETTINGS_MODULE lms.envs.production
+ENV DJANGO_SETTINGS_MODULE="lms.envs.$EDX_PLATFORM_SETTINGS"
 EXPOSE 8000
-CMD gunicorn -c /edx/app/edxapp/edx-platform/lms/docker_lms_gunicorn.py --name lms --bind=0.0.0.0:8000 --max-requests=1000 --access-logfile - lms.wsgi:application
+CMD gunicorn \
+    -c /edx/app/edxapp/edx-platform/lms/docker_lms_gunicorn.py \
+    --name lms \
+    --bind=0.0.0.0:8000 \
+    --max-requests=1000 \
+    --access-logfile \
+    - lms.wsgi:application
 
-FROM lms as lms-newrelic
-RUN pip install newrelic
-CMD newrelic-admin run-program gunicorn -c /edx/app/edxapp/edx-platform/lms/docker_lms_gunicorn.py --name lms --bind=0.0.0.0:8000 --max-requests=1000 --access-logfile - lms.wsgi:application
```

Comment:
tl;dr; I support this change, but putting my thinking down below in case it's useful to others..

At first, I had mixed feelings about this because I know that edX will need this bit for the image to be usable for them in production use-cases. Not having this here would require to create custom dockerfiles that built on top of this base image.  However, as I thought more about it, I realized there are other issues specific to edX that would require them to do that anyway.  In particular, they would need a custom image to install XBlocks and Plugins specific to edX.org that are not public or default for the Open edX Platform.

Reply:
FWIW, edX does have a repo for private Dockerfiles, in which they extend the public in-repo Dockerfile and change the gunicorn command to a newrelic command. They use this pattern already for at least [notes](https://github.com/edx/edx-notes-api/blob/master/Dockerfile#L750).